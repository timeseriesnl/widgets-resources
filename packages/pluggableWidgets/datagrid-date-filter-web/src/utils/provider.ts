import { Context, Dispatch } from "react";
import { ListAttributeValue } from "mendix";
import { FilterCondition } from "mendix/filters";

export interface FilterFunction {
    getFilterCondition(): FilterCondition | undefined;
}

export interface FilterContextValue {
    filterDispatcher: Dispatch<FilterFunction>;
    attribute: ListAttributeValue;
}

export function getFilterDispatcher(): Context<FilterContextValue> | undefined {
    return (window as any)["com.mendix.widgets.web.datagrid.filterContext"] as Context<FilterContextValue>;
}
