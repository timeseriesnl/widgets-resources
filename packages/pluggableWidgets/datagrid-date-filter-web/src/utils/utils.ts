/**
 * Change the time of a date and return an UTC date
 * @param date
 * @param hours
 * @param minutes
 * @param seconds
 */
export function changeTimeToMidnight(date: Date): Date {
    const newDate = new Date(date.getTime());
    newDate.setHours(0);
    newDate.setMinutes(0);
    newDate.setSeconds(0);
    newDate.setMilliseconds(0);
    return newDate;
}
