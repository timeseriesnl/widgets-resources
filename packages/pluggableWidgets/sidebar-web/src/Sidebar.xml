<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.web.sidebar.Sidebar" needsEntityContext="true" offlineCapable="true"
        pluginWidget="true"
        xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Sidebar</name>
    <description>Display sidebar content conditionally.</description>
    <icon/>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="contents" type="widgets" required="false">
                    <caption>Sidebar contents</caption>
                    <description/>
                </property>
                <property key="toggleMode" type="enumeration" defaultValue="push">
                    <caption>Toggle mode</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="none">None</enumerationValue>
                        <enumerationValue key="push">Push content aside</enumerationValue>
                        <enumerationValue key="slideOver">Slide over content</enumerationValue>
                        <enumerationValue key="startExpandedShrink">Shrink content (initially expanded)
                        </enumerationValue>
                        <enumerationValue key="startCollapsedShrink">Shrink content (initially collapsed)
                        </enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
            <propertyGroup caption="Dimensions">
                <property key="widthUnit" type="enumeration" defaultValue="pixels">
                    <caption>Width unit</caption>
                    <description>A percentage is relative to the viewport width.</description>
                    <enumerationValues>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                        <enumerationValue key="percentage">Percentage</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="widthValue" type="integer" defaultValue="100">
                    <caption>Width</caption>
                    <description/>
                </property>
                <property key="expandedWidthUnit" type="enumeration" defaultValue="pixels">
                    <caption>Expanded width unit</caption>
                    <description>A percentage is relative to the viewport width.</description>
                    <enumerationValues>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                        <enumerationValue key="percentage">Percentage</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="expandedWidthValue" type="integer" defaultValue="100">
                    <caption>Expanded width</caption>
                    <description/>
                </property>
                <property key="collapsedWidthUnit" type="enumeration" defaultValue="pixels">
                    <caption>Collapsed width unit</caption>
                    <description>A percentage is relative to the viewport width.</description>
                    <enumerationValues>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                        <enumerationValue key="percentage">Percentage</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="collapsedWidthValue" type="integer" defaultValue="100">
                    <caption>Collapsed width</caption>
                    <description/>
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
