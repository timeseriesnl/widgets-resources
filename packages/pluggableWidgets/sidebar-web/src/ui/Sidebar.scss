.widget-sidebar {
    background-color: #24276c;
    height: 100%;
    width: 0;
    transition: width 0.2s ease-in;
}

.widget-sidebar-slide-over {
    overflow: hidden;
    position: absolute;
    z-index: 999;
}
