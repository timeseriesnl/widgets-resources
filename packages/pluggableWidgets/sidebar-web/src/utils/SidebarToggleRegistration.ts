export function registerSidebarToggle(toggle: () => void): () => void {
    const sidebarToggleLocation = "com.mendix.widgets.web.sidebar.toggle";

    if ((window as any)[sidebarToggleLocation]) {
        throw Error(
            "Multiple sidebar widgets on a page isn't supported. Please model your page in such a way that there is just one sidebar."
        );
    }

    (window as any)[sidebarToggleLocation] = toggle;

    return () => ((window as any)[sidebarToggleLocation] = undefined);
}
