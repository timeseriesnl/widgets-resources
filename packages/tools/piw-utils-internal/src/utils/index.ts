export * from "./dom";
export * from "./PageEditorUtils";
export * from "./DataGridFilterIcons";
export * from "./StylingPropertyUtils";
