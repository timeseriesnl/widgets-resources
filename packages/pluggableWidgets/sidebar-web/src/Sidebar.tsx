import { createElement, ReactElement } from "react";

import { Sidebar as SidebarComponent } from "./components/Sidebar";

import { SidebarContainerProps } from "../typings/SidebarProps";

export function Sidebar(props: SidebarContainerProps): ReactElement | null {
    const width = getWidth(props.widthUnit, props.widthValue);
    const collapsedWidth = getWidth(props.collapsedWidthUnit, props.collapsedWidthValue);
    const expandedWidth = getWidth(props.expandedWidthUnit, props.expandedWidthValue);

    return (
        <SidebarComponent
            className={props.class}
            name={props.name}
            style={props.style}
            tabIndex={props.tabIndex}
            collapsible={props.toggleMode !== "none"}
            startExpanded={props.toggleMode === "none" || props.toggleMode === "startExpandedShrink"}
            width={props.toggleMode === "none" ? width : undefined}
            collapsedWidth={
                props.toggleMode === "startCollapsedShrink" || props.toggleMode === "startExpandedShrink"
                    ? collapsedWidth
                    : undefined
            }
            expandedWidth={expandedWidth}
            slideOver={props.toggleMode === "slideOver"}
        >
            {props.contents}
        </SidebarComponent>
    );
}

function getWidth(widthUnit: "pixels" | "percentage", width: number): string {
    return `${width}${widthUnit === "pixels" ? "px" : "vw"}`;
}
