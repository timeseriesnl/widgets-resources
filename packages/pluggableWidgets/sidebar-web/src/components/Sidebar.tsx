import { createElement, CSSProperties, PropsWithChildren, ReactElement, useEffect, useState } from "react";
import classNames from "classnames";

import "../ui/Sidebar.scss";
import { registerSidebarToggle } from "../utils/SidebarToggleRegistration";
import { Alert } from "@mendix/piw-utils-internal";

export interface SidebarProps {
    name: string;
    className?: string;
    style?: CSSProperties;
    tabIndex?: number;
    collapsible?: boolean;
    startExpanded?: boolean;
    width?: CSSProperties["width"];
    collapsedWidth?: CSSProperties["width"];
    expandedWidth?: CSSProperties["width"];
    slideOver?: boolean;
}

export function Sidebar(props: PropsWithChildren<SidebarProps>): ReactElement {
    const [expanded, setExpanded] = useState(props.startExpanded);
    const [error, setError] = useState("");

    let width: string | number | undefined;

    if (props.collapsible) {
        width = expanded ? props.expandedWidth : props.collapsedWidth;
    } else {
        width = props.width;
    }

    useEffect(() => {
        if (props.collapsible) {
            let unregisterSidebarToggle: () => void;

            try {
                unregisterSidebarToggle = registerSidebarToggle(() => setExpanded(prevExpanded => !prevExpanded));
            } catch (e) {
                setError(e.message);
            }

            return () => unregisterSidebarToggle();
        }
    }, [props.name, props.collapsible]);

    if (error) {
        return <Alert bootstrapStyle={"danger"}>{error}</Alert>;
    }

    return (
        <aside
            className={classNames(
                "widget-sidebar",
                {
                    "widget-sidebar-slide-over": props.slideOver,
                    "widget-sidebar-expanded": props.collapsible && expanded
                },
                props.className
            )}
            style={{ ...props.style, width }}
            tabIndex={props.tabIndex}
        >
            {props.children}
        </aside>
    );
}
