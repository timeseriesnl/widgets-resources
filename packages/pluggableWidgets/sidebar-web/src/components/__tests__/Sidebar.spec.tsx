import { createElement } from "react";
import { mount } from "enzyme";

import { Sidebar, SidebarProps } from "../Sidebar";

jest.mock("../../utils/SidebarToggleRegistration");
import { registerSidebarToggle as mockedRegisterSidebarToggle } from "../../utils/SidebarToggleRegistration";
import { act } from "react-dom/test-utils";
const { registerSidebarToggle } = jest.requireActual("../../utils/SidebarToggleRegistration");

(mockedRegisterSidebarToggle as any).mockImplementation((toggle: () => void) => registerSidebarToggle(toggle));

describe("Sidebar", () => {
    let defaultSidebarProps: SidebarProps;

    beforeEach(() => {
        defaultSidebarProps = {
            name: "Sidebar",
            className: "sidebar-left",
            style: {
                backgroundColor: "blue"
            },
            tabIndex: 1,
            collapsible: true,
            startExpanded: false,
            expandedWidth: 400
        };
    });

    afterEach(() => {
        (window as any)["com.mendix.widgets.web.sidebar.toggle"] = undefined;
    });

    it("renders starting collapsed", () => {
        const SidebarWrapper = mount(<Sidebar {...defaultSidebarProps} />);

        expect(SidebarWrapper.getDOMNode()).toMatchSnapshot();
    });

    it("renders starting expanded", () => {
        const SidebarWrapper = mount(<Sidebar {...defaultSidebarProps} startExpanded />);

        expect(SidebarWrapper.getDOMNode()).toMatchSnapshot();
    });

    it("renders with slide over mode", () => {
        const SidebarWrapper = mount(<Sidebar {...defaultSidebarProps} slideOver />);

        expect(SidebarWrapper.getDOMNode()).toMatchSnapshot();
    });

    it("renders with collapsed width", () => {
        const SidebarWrapper = mount(<Sidebar {...defaultSidebarProps} collapsedWidth={100} expandedWidth={500} />);

        expect(SidebarWrapper.getDOMNode()).toMatchSnapshot();
    });

    it("renders an alert in the case of a failure", () => {
        (mockedRegisterSidebarToggle as any).mockImplementationOnce(() => {
            throw Error("Error message");
        });
        const SidebarWrapper = mount(<Sidebar {...defaultSidebarProps} />);

        expect(SidebarWrapper.getDOMNode()).toMatchSnapshot();
    });

    it("is toggleable when collapsible", () => {
        const sidebarWrapper = mount(<Sidebar {...defaultSidebarProps} />);

        expect((sidebarWrapper.find("aside").getDOMNode() as HTMLElement).className).not.toContain(
            "widget-sidebar-expanded"
        );

        act(() => {
            (window as any)["com.mendix.widgets.web.sidebar.toggle"]?.();
        });

        sidebarWrapper.update();
        expect((sidebarWrapper.find("aside").getDOMNode() as HTMLElement).className).toContain(
            "widget-sidebar-expanded"
        );
        expect((sidebarWrapper.find("aside").getDOMNode() as HTMLElement).style.width).toBe("400px");
    });

    it("isn't toggleable when not collapsible", () => {
        const sidebarWrapper = mount(
            <Sidebar {...defaultSidebarProps} collapsible={false} expandedWidth={undefined} width={500} />
        );

        expect((sidebarWrapper.find("aside").getDOMNode() as HTMLElement).className).not.toContain(
            "widget-sidebar-expanded"
        );
        expect((sidebarWrapper.find("aside").getDOMNode() as HTMLElement).style.width).toBe("500px");

        act(() => {
            (window as any)["com.mendix.widgets.web.sidebar.toggle"]?.();
        });

        sidebarWrapper.update();
        expect((sidebarWrapper.find("aside").getDOMNode() as HTMLElement).className).not.toContain(
            "widget-sidebar-expanded"
        );
        expect((sidebarWrapper.find("aside").getDOMNode() as HTMLElement).style.width).toBe("500px");
    });

    // TODO: enable this test after upgrading the enzyme-adapter-react-17 dep to the latest version, since this will call the useEffect cleanup function
    xit("cleans up", () => {
        const sidebarWrapper = mount(<Sidebar {...defaultSidebarProps} />);

        expect((window as any)["com.mendix.widgets.web.sidebar.toggle"]).toBeTruthy();

        sidebarWrapper.unmount();
        expect((window as any)["com.mendix.widgets.web.sidebar.toggle"]).toBeUndefined();
    });
});
