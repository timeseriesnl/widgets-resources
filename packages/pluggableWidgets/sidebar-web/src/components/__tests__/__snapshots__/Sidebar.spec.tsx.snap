// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders an alert in the case of a failure 1`] = `
<div
  class="alert alert-danger"
>
  Error message
</div>
`;

exports[`Sidebar renders starting collapsed 1`] = `
<aside
  class="widget-sidebar sidebar-left"
  style="background-color: blue;"
  tabindex="1"
/>
`;

exports[`Sidebar renders starting expanded 1`] = `
<aside
  class="widget-sidebar widget-sidebar-expanded sidebar-left"
  style="background-color: blue; width: 400px;"
  tabindex="1"
/>
`;

exports[`Sidebar renders with collapsed width 1`] = `
<aside
  class="widget-sidebar sidebar-left"
  style="background-color: blue; width: 100px;"
  tabindex="1"
/>
`;

exports[`Sidebar renders with slide over mode 1`] = `
<aside
  class="widget-sidebar widget-sidebar-slide-over sidebar-left"
  style="background-color: blue;"
  tabindex="1"
/>
`;
